pkg/js/validator/validate-store.ts

```diff
@@ -0,0 +1,976 @@
+          errors.push(invalidTypeUser(listType, types, instancePath + `/user_filter/${typeNo}/type`));
+        }
+      }
+
+      // Check relations if present
+      const relation = listUsers.user_filter[typeNo].relation;
+      if (relation && isStringValue(relation)) {
+        const typeRelations = model.type_definitions.filter((rel) => rel.type === listType).map((rel) => rel.relations);
+
+        if (typeRelations.length && typeRelations[0] && Object.keys(typeRelations[0]).length) {
+          if (!Object.keys(typeRelations[0]).includes(relation)) {
+            errors.push(
+              invalidRelationUser(
+                relation,
+                Object.keys(typeRelations[0]),
+                instancePath + `/user_filter/${typeNo}/relation`,
+              ),
+            );
+          }
+        } else {
+          errors.push(relationMustExistOnType(relation, listType, instancePath + `/user_filter/${typeNo}/relation`));
+        }
+      }
+    }
+
+    // Check assertions
+    if (listUsers.assertions) {
+      for (const assertion of Object.keys(listUsers.assertions)) {
+        if (listUsers.assertions[assertion].users) {
+          for (const user of listUsers.assertions[assertion].users) {
+            if (!tuples.some((tuple) => tuple.user === user)) {
+              errors.push(undefinedTypeTuple(user, instancePath + `/assertions/${assertion}/users`));
+            }
+          }
+        }
+      }
+    }
+  }
+
+  // Check context params
+  if (listUsers.context) {
+    for (const testParam in listUsers.context) {
+      if (!params.includes(testParam)) {
+        errors.push(unidentifiedTestParam(testParam, instancePath + `/context/${testParam}`));
+      }
+    }
+  }
+
+  return errors;
+}
+
+// Validation for types in check
+function validateTestTypes(store: Store, model: AuthorizationModel, instancePath: string): boolean {
+  const errors = [];
+
+  // Collect params for validity checking
+  const params: string[] = [];
+  for (const condition in model.conditions) {
+    for (const param in model.conditions[condition].parameters) {
+      params.push(param);
+    }
+  }
+
+  for (const testNo in store.tests) {
+    // Collect valid tuples
+    const tuples = [];
+    if (store.tuples && store.tuples.length) {
+      tuples.push(...store.tuples);
+    }
+
+    const test = store.tests[testNo];
+    if (!test) {
+      continue;
+    }
+
+    if (test.tuples && test.tuples.length) {
+      tuples.push(...test.tuples);
+    }
+
+    // Validate check
+    for (const checkNo in test.check) {
+      if (!test.check[checkNo] || !test.check[checkNo].user || !test.check[checkNo].object) {
+        return false;
+      }
+      errors.push(
+        ...validateCheck(
+          model,
+          test.check[checkNo],
+          tuples,
+          params,
+          instancePath + `/tests/${testNo}/check/${checkNo}`,
+        ),
+      );
+    }
+
+    // Validate list objects
+    for (const listNo in test.list_objects) {
```

Comment:
```suggestion
    for (const singleListObjectsTest of test.list_objects) {
```

Reply:
That way you don't need to do `if !test.list_objects[listNo]` below, just
```
if !singleListObjectsTest.user || !singleListObjectsTest.type)
```

Reply:
Problem with `of` is that it throws a `is not iterable` exception on **undefined** values, where as `in` doesn't.

Happy to make changes to use `of`, I'll just have to null-check ahead of the loop - is that preferable for you?